crates/iota-core/src/transaction_manager.rs

```diff
@@ -438,15 +438,34 @@ impl TransactionManager {
         let mut receiving_objects: HashSet<InputKey> = HashSet::new();
         let certs: Vec<_> = certs
             .into_iter()
-            .map(|(cert, fx_digest)| {
+            .filter_map(|(cert, fx_digest)| {
                 let input_object_kinds = cert
                     .data()
                     .intent_message()
                     .value
                     .input_objects()
                     .expect("input_objects() cannot fail");
                 let mut input_object_keys =
-                    epoch_store.get_input_object_keys(&cert.key(), &input_object_kinds);
+                    match epoch_store.get_input_object_keys(&cert.key(), &input_object_kinds) {
+                        Ok(keys) => keys,
+                        Err(e) => {
+                            // Because we do not hold the transaction lock during enqueue, it is
+                            // possible that the transaction was
+                            // executed and the shared version assignments deleted
+                            // since the earlier check. This is a rare race condition, and it is
+                            // better to handle it ad-hoc here than to
+                            // hold tx locks for every cert for the duration
+                            // of this function in order to remove the race.
+                            if self
+                                .transaction_cache_read
+                                .is_tx_already_executed(cert.digest())
+                                .expect("is_tx_already_executed cannot fail")
+                            {
+                                return None;
+                            }
+                            panic!("Failed to get input object keys: {:?}", e);
```

Comment:
Why are we using panics instead of propagating errors?

Reply:
I think this should not happen?

Reply:
I guess this is just how this code works 🤷‍♀️ as there seems to be no other error propagation